### PR TITLE
Programming guide update

### PR DESCRIPTION
**Pull request by hadarhg**
- Added data encoding section for Java and Python
- Changed some Python code snippets to use include tags

### CODE REVIEW

**Comment by davorbonaci** [comment]
R: @francesperry

File: src/documentation/programming-guide.md [hunk 1]
```diff
@@ -174,20 +168,8 @@ public static void main(String[] args) {
...
-  "Whether 'tis nobler in the mind to suffer ",
-  "The slings and arrows of outrageous fortune, ",
-  "Or to take arms against a sea of troubles, "
-]
-
-# Create the pipeline.
-p = beam.Pipeline()
-
-collection = p | 'ReadMyLines' >> beam.Create(lines)
+% github_sample /apache/beam/blob/master/sdks/python/apache_beam/examples/snippets/snippets.py tag:model_pcollection
```
Comment (francesperry) [inline src/documentation/programming-guide.md]: Worth doing a quick diff of the generated content to make sure these changes look as expected?

File: src/documentation/programming-guide.md [hunk 2]
```diff
@@ -1079,13 +1052,185 @@ For blocking execution, append the `waitUntilFinish` method:
...
+
+#### Default coders and the CoderRegistry
+
+Each Pipeline object has a `CoderRegistry` object, which maps language types to the default coder the pipeline should use for those types. You can use the `CoderRegistry` yourself to look up the default coder for a given type, or to register a new default coder for a given type.
+
+{:.language-java}
+`CoderRegistry` contains a default mapping of coders to standard Java types for any Pipeline you create using the Beam SDK for Java. The following table shows the standard mapping:
+
+{:.language-java}
+| Java Type        | Default Coder         |
```
Comment (francesperry) [inline src/documentation/programming-guide.md]: We're using different table formats across the site. Is there any easy way to make this match?

File: src/documentation/programming-guide.md [hunk 3]
```diff
@@ -1079,13 +1052,185 @@ For blocking execution, append the `waitUntilFinish` method:
...
+
+
+
+##### Looking up a default coder
+
+{:.language-java}
+You can use the method CoderRegistry.getDefaultCoder to determine the default Coder for a Java type. You can access the CoderRegistry for a given Pipeline by using the method Pipeline.getCoderRegistry. This allows you to determine (or set) the default Coder for a Java type on a per-pipeline basis: i.e. "for this pipeline, verify that Integer values are encoded using BigEndianIntegerCoder."
+
+{:.language-py}
+You can use the method CoderRegistry.get_coder to determine the default Coder for a Python type. You can use `coders.registry` to access the CoderRegistry. This allows you to determine (or set) the default Coder for a Python type on a per-pipeline basis.
```
Comment (francesperry) [inline src/documentation/programming-guide.md]: Is there a reason this toggles at the paragraph level?  Seems like more text could be shared and simplify the markdown.

File: src/documentation/programming-guide.md [hunk 4]
```diff
@@ -1079,13 +1052,185 @@ For blocking execution, append the `waitUntilFinish` method:
...
+| Long             | VarLongCoder          |
+| String           | StringUtf8Coder       |
+| TableRow         | TableRowJsonCoder     |
+| Void             | VoidCoder             |
+| byte[ ]          | ByteArrayCoder        |
+| TimestampedValue | TimestampedValueCoder |
+
+
+{:.language-py}
+`CoderRegistry` contains a default mapping of Coders to standard Python types for any Pipeline you create using the Beam SDK for Python. The following table shows the standard mapping:
```
Comment (francesperry) [inline src/documentation/programming-guide.md]: Can this paragraph be unified with Java one?

File: src/documentation/programming-guide.md (hunk 3 above)
Comment (francesperry) [inline src/documentation/programming-guide.md]: code font around 
CoderRegistry.get_coder

File: src/documentation/programming-guide.md [hunk 5]
```diff
@@ -1079,13 +1052,185 @@ For blocking execution, append the `waitUntilFinish` method:
...
+| str         | BytesCoder    |
+| bytes       | StrUtf8Coder  |
+| Tuple       | TupleCoder    |
+
+
+
+##### Looking up a default coder
+
+{:.language-java}
+You can use the method CoderRegistry.getDefaultCoder to determine the default Coder for a Java type. You can access the CoderRegistry for a given Pipeline by using the method Pipeline.getCoderRegistry. This allows you to determine (or set) the default Coder for a Java type on a per-pipeline basis: i.e. "for this pipeline, verify that Integer values are encoded using BigEndianIntegerCoder."
```
Comment (francesperry) [inline src/documentation/programming-guide.md]: Code font on CoderRegistry.getDefaultCoder

Comment (francesperry) [inline src/documentation/programming-guide.md]: verify that -> use?

(And is there a reason Py doesn't have the same example?)

File: src/documentation/programming-guide.md [hunk 6]
```diff
@@ -1079,13 +1052,185 @@ For blocking execution, append the `waitUntilFinish` method:
...
+##### Setting the default coder for a type
+
+{:.language-java}
+To set the default Coder for a Java type for a particular pipeline, you obtain and modify the pipeline's `CoderRegistry`. You use the method `Pipeline.getCoderRegistry` to get the `CoderRegistry` object, and then use the method `CoderRegistry.registerCoder` to register a new `Coder` for the target Java type.
+
+{:.language-java}
+The following example code demonstrates how to set a default Coder, in this case `BigEndianIntegerCoder`, for Integer values for a pipeline.
+
+{:.language-py}
+To set the default Coder for a Python type, you obtain and modify the `CoderRegistry`. You may use the `coders.registry` to get the `CoderRegistry` object, and then use the method `CoderRegistry.register_coder` to register a new Coder for the target Python type.
```
Comment (francesperry) [inline src/documentation/programming-guide.md]: Seems like another example of a paragraph that might be able to be written once with only the method names toggling?

File: src/documentation/programming-guide.md (hunk 5 above)
Comment (hadarhg) [inline src/documentation/programming-guide.md]: I changed the Python version a bit. According to Ahmet, this is how it is for the Python SDK.

**Pull request closed without merging**